contracts/oracle/src/contract/alarms.rs

```diff
@@ -1,62 +1,216 @@
+            addr: Addr,
+            above: Option<SpotPrice>,
+            price_alarms: PriceAlarms<'m>,
+        }
+
+        impl<'m, BaseC> WithQuote<BaseC> for AddAlarms<'m>
+        where
+            BaseC: Currency,
+        {
+            type Output = Response;
+            type Error = ContractError;
+
+            fn exec<C>(self, below: Price<C, BaseC>) -> Result<Self::Output, Self::Error>
+            where
+                C: Currency,
+            {
+                if let Some(above) = self.above {
+                    self.price_alarms.add_alarm_above::<C, BaseC>(
+                        self.storage,
+                        &self.addr,
+                        above.try_into()?,
+                    )?;
+                }
+                self.price_alarms
+                    .add_alarm_below(self.storage, &self.addr, below)?;
+                Ok(Response::new())
+            }
+        }
+
+        let (below, above) = alarm.into();
+        with_quote::execute::<_, _, _, BaseC>(
+            &below,
+            AddAlarms {
+                storage,
+                addr,
+                above,
+                price_alarms: Self::PRICE_ALARMS,
+            },
+        )
+    }
+
+    fn schedule_alarm(
+        batch: &mut Batch,
+        addr: Addr,
+        next_id: &mut AlarmReplyId,
+    ) -> Result<(), ContractError> {
+        // TODO: get rid of the Nls dummy type argument
+        batch.schedule_execute_wasm_reply_always::<_, Nls>(
+            &addr,
+            ExecuteAlarmMsg::PriceAlarm(),
+            None,
+            *next_id,
+        )?;
+
+        *next_id = next_id.wrapping_add(1);
+
+        Ok(())
     }
 
-    pub fn try_notify_alarms(
+    pub fn try_notify_alarms<BaseC>(
         storage: &mut dyn Storage,
         mut batch: Batch,
         prices: &[SpotPrice],
-        max_count: u32,
-    ) -> Result<Response, ContractError> {
-        let sent = Self::PRICE_ALARMS.notify(storage, &mut batch, prices, max_count)?;
-        Ok(Response::from(batch).set_data(to_binary(&DispatchAlarmsResponse(sent))?))
+        max_count: u32, // TODO: type alias
+    ) -> Result<Response, ContractError>
+    where
+        BaseC: Currency,
+    {
+        struct NotifyCmd<'a> {
+            storage: &'a dyn Storage,
+            price_alarms: &'a PriceAlarms<'static>,
+            max_count: u32,
+            next_id: &'a mut AlarmReplyId,
+            batch: &'a mut Batch,
+        }
+
+        impl<'a, BaseC> WithQuote<BaseC> for NotifyCmd<'a>
+        where
+            BaseC: Currency,
+        {
+            type Error = ContractError;
+            type Output = u32;
+
+            fn exec<C>(self, price: Price<C, BaseC>) -> Result<Self::Output, Self::Error>
+            where
+                C: Currency,
+            {
+                let initial = *self.next_id;
+
+                self.price_alarms
+                    .alarms(self.storage, price)
+                    .take(self.max_count as usize)
```

Comment:
please avoid using `as` since it may mask overflow issues, try_into()? would be a better alternative. I realize overflow may never happen since `usize` at wasm32 should be u32 ... but this builds bad code practices.